seqfold/seqfold.go

```diff
@@ -0,0 +1,157 @@
+	"math"
+	"strings"
+
+	"github.com/TimothyStiles/poly/checks"
+)
+
+type Comp map[byte]byte
+
+type MultiBranchEnergies struct {
+	A, B, C, D float64
+}
+
+// Energy holds two energies, enthaply and entropy
+// SantaLucia & Hicks (2004), Annu. Rev. Biophys. Biomol. Struct 33: 415-440
+type Energy struct {
+	// enthaply
+	H float64
+	// entropy
+	S float64
+}
+
+type BpEnergy map[string]Energy
+
+type LoopEnergy map[int]Energy
+
+type Energies struct {
+	BULGE_LOOPS     LoopEnergy
+	COMPLEMENT      Comp
+	DE              BpEnergy
+	HAIRPIN_LOOPS   LoopEnergy
+	MULTIBRANCH     MultiBranchEnergies
+	INTERNAL_LOOPS  LoopEnergy
+	INTERNAL_MM     BpEnergy
+	NN              BpEnergy
+	TERMINAL_MM     BpEnergy
+	TRI_TETRA_LOOPS BpEnergy
+}
+
+type Interval struct {
+	I, J int
+}
+
+// A single structure with a free energy, description, and inward children
+type NucleicAcidStructure struct {
+	// Desc is the description of the NucleicAcidStructure
+	Desc string
+	// Inner is list of inner structure represented as intervals in the
+	// sequence
+	Inner []Interval
+	// E is the energy of the NucleicAcidStructure
+	E float64
+}
+
+func (s NucleicAcidStructure) Equal(other NucleicAcidStructure) bool {
+	if len(s.Inner) != len(other.Inner) {
+		return false
+	}
+	for i, val := range s.Inner {
+		if val != other.Inner[i] {
+			return false
+		}
+	}
+	return s.E == other.E
+}
+
+func (s NucleicAcidStructure) Valid() bool {
+	return s.E != math.Inf(1) && s.E != math.Inf(-1)
+}
+
+func (s NucleicAcidStructure) String() string {
+	i, j := "", ""
+	if len(s.Inner) > 0 {
+		i, j = fmt.Sprint(s.Inner[0].I), fmt.Sprint(s.Inner[0].J)
+	}
+	return fmt.Sprintf("%4s %4s % 6.2f  %-15s", i, j, s.E, s.Desc)
+}
+
+// MultiString returns all the fields as strings, this is useful to output the
+// result in a tabular fashion using the same format string.
+func (s NucleicAcidStructure) MultiString() (string, string, string, string) {
+	i, j := "", ""
+	if len(s.Inner) > 0 {
+		i, j = fmt.Sprint(s.Inner[0].I), fmt.Sprint(s.Inner[0].J)
+	}
+	return i, j, fmt.Sprintf("%6.2f", s.E), s.Desc
+}
+
+// STRUCT_DEFAULT = Struct(-math.inf)
+var STRUCT_DEFAULT = NucleicAcidStructure{
+	Desc: "",
+	E:    math.Inf(-1),
+}
+
+// STRUCT_NULL = Struct(math.inf)
+var STRUCT_NULL = NucleicAcidStructure{
+	Desc: "",
+	E:    math.Inf(1),
+}
+
+// FoldContext holds the energy caches, energy maps, sequence, and temperature
+// needed in order to compute the folding energy and structures.
+type FoldContext struct {
+	Energies Energies
+	Seq      string
+	V        [][]NucleicAcidStructure
+	W        [][]NucleicAcidStructure
+	T        float64
+}
+
+// NewFoldingContext returns a FoldContext ready to use.
+func NewFoldingContext(seq string, temp float64) (FoldContext, error) {
+	seq = strings.ToUpper(seq)
+
+	// figure out whether it's DNA or RNA, choose energy map
+	var emap Energies
+	switch {
+	case checks.IsDNA(seq):
+		emap = DNA_ENERGIES
+	case checks.IsRNA(seq):
+		emap = RNA_ENERGIES
+	default:
```

Comment:
Tired so maybe this logic is already solved but what happens when a sequence contains neither T or U?

Reply:
It is considered DNA since it is the first to be checked.
Not much else we can do in this case IMHO...